File: crates/yokan/demo/edges.py
# /// script
# requires-python = ">=3.14"
# ///
"""Containment, proven by the gate: every predictable failure an
admitted program can reach fails the SAME way interpreted and
compiled — the statement aborts before writing, earlier statements'
effects stay, the app keeps running.

  oob     — a local subscript past the end
  grow    — i64 overflow (the write is refused before it happens,
            in both runs)
  partial — first statement lands, second fails: ordering agrees
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from yokan import button, column, run, State, text  # noqa: E402

xs: State[list[int]] = State([7])
picked: State[int] = State(0)
big: State[int] = State(4611686018427387904)
steps: State[int] = State(0)


def oob():
    r = xs()
    picked.set(r[5])


def grow():
    big.set(big() * 4)


def partial():
    steps.set(steps() + 1)
    r = xs()
    picked.set(r[9])


def view():
    with column(spacing=8, padding=12):
        text(f"picked={picked()} steps={steps()}")
        button("oob", on_click=oob)
        button("grow", on_click=grow)
        button("partial", on_click=partial)


if __name__ == "__main__":
    run(view, title="edges")

File: gomamochi/demo/edges.go
// The edges: an index past the end of a list, and a number far past
// what a machine word holds that keeps growing. Both runs have to
// answer the same, and this is the demo that says so.
//
// Go answers the first by refusing: an index past the end stops the
// program in both runs, so the app asks the length first. The second
// is `math/big`, the same package in both runs.
package main

import (
	"fmt"
	"math/big"

	. "github.com/i2y/yokan/gomamochi"
)

type Edges struct {
	xs     []int
	picked int
	big    *big.Int
	steps  int
}

// The element at i, or -1 when there is none.
func (e *Edges) at(i int) int {
	if i < 0 || i >= len(e.xs) {
		return -1
	}
	return e.xs[i]
}

func (e *Edges) View() Element {
	return Column(
		Text(fmt.Sprintf("picked=%d steps=%d", e.picked, e.steps)),
		Text("big="+e.big.String()),
		Button("oob").OnClick(func() { e.picked = e.at(5) }),
		Button("grow").OnClick(func() { e.big = new(big.Int).Mul(e.big, big.NewInt(4)) }),
		Button("partial").OnClick(func() {
			e.steps += 1
			e.picked = e.at(9)
		}),
	).Spacing(8).Padding(12)
}

func main() {
	start, _ := new(big.Int).SetString("18446744073709551616", 10)
	Run(&Edges{xs: []int{7}, big: start}, Title("edges"))
}
